src/transformers/modeling_utils.py

```diff
@@ -849,21 +846,40 @@ def generate(
 
         if self.config.is_encoder_decoder:
             assert bos_token_id is not None, "Encoder Decoder Models need to have a bos_token_id"
-            # encoder decoder need to start with empty input_ids and copy the input_ids to encoder_inputs
-            encoder_inputs = input_ids
+
+            # only need to generate encoder_outputs once for encoder-decoder
+            if hasattr(self, "encoder"):
```

Comment:
fairseq has models implement a method called `encode_forward` and then the various generators just call that. Will get rid of these ifs.

Reply:
Yeah, I agree this is not ideal here. I mean BartModel would never `generate()` because it does not have an `lm_head`, but I think you're right this logic should be in each of the different model files. So we could introduce a standard for encoder-decoder with `lm_head` models to have a `encode_forward` function. I would be fine with that @thomwolf @craffel.

Reply:
changed it. What do you think @sshleifer ?